tests/suites/test_suite_pk.function

```diff
@@ -526,9 +527,23 @@ void pk_rsa_verify_ext_test_vec( data_t * message_str, int digest,
         pss_opts.expected_salt_len = salt_len;
     }
 
-    TEST_ASSERT( mbedtls_pk_verify_ext( pk_type, options, &pk,
-                                digest, hash_result, hash_len,
-                                result_str->x, sig_len ) == result );
+    ret = mbedtls_pk_verify_ext( pk_type, options, &pk,
+                                 digest, hash_result, hash_len,
+                                 result_str->x, sig_len );
+
+    /* Mbed TLS distinguishes "invalid padding" from "valid padding but
+     * the rest of the signature is invalid". This has little use in
+     * practice and PSA doesn't report this distinction.
+     * In this case, PSA returns PSA_ERROR_INVALID_SIGNATURE translated
+     * to MBEDTLS_ERR_RSA_VERIFY_FAILED
+     */
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    if( result == MBEDTLS_ERR_RSA_INVALID_PADDING &&
+        ret == MBEDTLS_ERR_RSA_VERIFY_FAILED )
+        TEST_EQUAL( ret, MBEDTLS_ERR_RSA_VERIFY_FAILED);
+    else
+#endif
+    TEST_EQUAL( ret, result );
```

Comment:
This block of code looks rather weird. If `ret` is `VERIFY_FAILED`, we assert that `ret` is `VERIFY_FAILED`. The last line has the wrong indentation if `defined(MBEDTLS_USE_PSA_CRYPTO)` — normally we'd resolve this with braces around it so that the presence of the `else` doesn't change the expected indentation.

Here's how I'd write it.
```suggestion
#if defined(MBEDTLS_USE_PSA_CRYPTO)
    if( result == MBEDTLS_ERR_RSA_INVALID_PADDING )
        result = MBEDTLS_ERR_RSA_VERIFY_FAILED;
#endif
    TEST_EQUAL( ret, result );
```
We've followed this pattern in a few other places, e.g. https://github.com/Mbed-TLS/mbedtls/blob/d65aeb37349ad1a50e0f6c9b694d4b5290d60e49/tests/suites/test_suite_cipher.function#L613

Reply:
I would have done that if it was possible, but it's not because INVALID_PADDING can be returned for other conditions, I already replied to @mpg's review at https://github.com/Mbed-TLS/mbedtls/pull/5742#discussion_r854962386

Maybe the following would be ok ?
```suggestion
#if defined(MBEDTLS_USE_PSA_CRYPTO)
    if( result == MBEDTLS_ERR_RSA_INVALID_PADDING &&
        ret == MBEDTLS_ERR_RSA_VERIFY_FAILED )
        result = MBEDTLS_ERR_RSA_VERIFY_FAILED;
#endif
    TEST_EQUAL( ret, result );
```

Reply:
A signature verification function returns `MBEDTLS_ERR_RSA_INVALID_PADDING` when `USE_PSA_CRYPTO` is active? This doesn't sound right: the underlying PSA function shouldn't return `PSA_ERROR_INVALID_PADDING`, and I'm not seeing code in `pk*.c` that translates other PSA errors to `MBEDTLS_ERR_RSA_INVALID_PADDING`. What am I missing? Is there a bug in PSA code that makes it return `PSA_ERROR_INVALID_PADDING` for a signature verification? I guess that would be plausible if `rsa.c` code is returning `MBEDTLS_ERR_RSA_INVALID_PADDING` which gets translated to `PSA_ERROR_INVALID_PADDING` by the generic error conversion function `mbedtls_to_psa_error` — but that would be a minor bug in the PSA code.

Reply:
It's returned by an mbedtls call to `mbedtls_rsa_rsassa_pss_verify_ext` 
https://github.com/Mbed-TLS/mbedtls/blob/6c242a01f75cedfd0f156d70f84a236c5b957cea/library/pk.c#L441-L445

It seems only some specific RSA-PSS verify crypto can be passed to PSA:
https://github.com/Mbed-TLS/mbedtls/blob/6c242a01f75cedfd0f156d70f84a236c5b957cea/library/pk.c#L377-L380

Reply:
I see, thanks for the explanation.

So please add a comment explaining this. And I'd still prefer to change the structure of the code to something more readable. How about this?
```suggestion
#if defined(MBEDTLS_USE_PSA_CRYPTO)
    if( result == MBEDTLS_ERR_RSA_INVALID_PADDING &&
        ret == MBEDTLS_ERR_RSA_VERIFY_FAILED )
    {
        // explanation goes here, no code needed
    }
    else
#endif
    {
        TEST_EQUAL( ret, result );
    }
```

Reply:
> INVALID_PADDING can be returned for other conditions

Oh, that's the bit I missed initially. I like Gilles's proposal for the code structure, and I suggest the explanation includes that point - without that I couldn't understand why the condition was what it is.

Reply:
Or maybe another way to express the same thing:
```c
#if defined(MBEDTLS_USE_PSA_CRYPTO)
    if( result == MBEDTLS_ERR_RSA_INVALID_PADDING )
    {
        /* may return either depending on which path was taken */
        TEST_ASSERT( ret == result || ret == MBEDTLS_ERR_RSA_VERIFY_FAILED );
    }
    else
#endif
    {
        TEST_EQUAL( ret, result );
    }
```
I think I would find that clearer.

Reply:
Thanks, this is what I'm ready to push based on your suggestions:
```
#if defined(MBEDTLS_USE_PSA_CRYPTO)
    if( result == MBEDTLS_ERR_RSA_INVALID_PADDING )
    {
        /* mbedtls_pk_verify_ext() may return MBEDTLS_ERR_RSA_INVALID_PADDING
         * error depending on which path was taken.
         * If the PSA path is used, it won't because Mbed TLS
         * distinguishes "invalid padding" from "valid padding but
         * the rest of the signature is invalid". This has little use in
         * practice and PSA doesn't report this distinction.
         * In this case, PSA returns PSA_ERROR_INVALID_SIGNATURE translated
         * to MBEDTLS_ERR_RSA_VERIFY_FAILED
         */
        TEST_ASSERT( ret == result || ret == MBEDTLS_ERR_RSA_VERIFY_FAILED );
    }
    else
#endif
    {
        TEST_EQUAL( ret, result );
    }
```